DAQ/src/StrawRecoFromArtdaqFragments_module.cc

```diff
@@ -0,0 +1,297 @@
+
+          int link_id = rdh->linkID;
+          TLOG(TLVL_DEBUG) << "DTC:" << dtc_id << " ROC:" << link_id
+                           << " offset:" << std::hex
+                           << " nhits:" << std::dec << nhits;
+
+          for (int ihit=0; ihit<nhits; ihit++) {
+//-----------------------------------------------------------------------------
+// first packet, 16 bytes, or 8 ushort's is the data header packet
+//-----------------------------------------------------------------------------
+            mu2e::TrackerDataDecoder::TrackerDataPacket* hit ;
+            int offset          = (ihit*np_per_hit+1)*packet_size;   // in bytes
+            hit = (mu2e::TrackerDataDecoder::TrackerDataPacket*) (roc_data+offset);
+//-----------------------------------------------------------------------------
+// unpack hit and fill the StrawDigiCollection
+// HV side has an extra bit in teh straw ID
+//-----------------------------------------------------------------------------
+            uint16_t straw_id = hit->StrawIndex;
+            if (straw_id >= 0x80) straw_id = straw_id - 0x80;
+
+            mu2e::StrawId sid(hit->StrawIndex);
+            mu2e::TrkTypes::TDCValues tdc = {hit->TDC0(), hit->TDC1()};
+            mu2e::TrkTypes::TOTValues tot = {hit->TOT0, hit->TOT1};
+            mu2e::TrkTypes::ADCValue  pmp = hit->PMP;
+
+            TLOG(TLVL_DEBUG+1) << "times:" << std::setw(10) << hit->TDC0() << " " << hit->TDC1()
+                               << " TOT:" << std::setw(4) << hit->TOT0 << " " << hit->TOT1
+                               << " pmp:" << hit->PMP;
+
+            straw_digis->emplace_back(sid, tdc, tot, pmp);
+//------------------------------------------------------------------------------
+// the corresponding waveform
+//-----------------------------------------------------------------------------
+            if (saveWaveforms_) {
+              std::vector<uint16_t> wf(nsamples);
+
+              wf[0] = hit->ADC00;
+              wf[1] = hit->ADC01();
+              wf[2] = hit->ADC02;
+
+              auto npackets   = 0;
+              auto idx        = 2;
+              auto adc_packet = (mu2e::TrackerDataDecoder::TrackerADCPacket*)((char*) hit + 16); // the packet size is 16 bytes
+              while (npackets < nADCPackets_) {
+                wf[++idx] = adc_packet->ADC0;
+                wf[++idx] = adc_packet->ADC1();
+                wf[++idx] = adc_packet->ADC2;
+                wf[++idx] = adc_packet->ADC3;
+                wf[++idx] = adc_packet->ADC4();
+                wf[++idx] = adc_packet->ADC5;
+                wf[++idx] = adc_packet->ADC6;
+                wf[++idx] = adc_packet->ADC7();
+                wf[++idx] = adc_packet->ADC8;
+                wf[++idx] = adc_packet->ADC9;
+                wf[++idx] = adc_packet->ADC10();
+                wf[++idx] = adc_packet->ADC11;
+                npackets++;
+                adc_packet++;
```

Comment:
The only general concern I have is that I don't see any integrity checks. It is an efficient deserialization assuming the fragments are error-free, but it may produce uncaught exceptions, or even terminate ungracefully, if there is an error in the payload. The canonical example being the missing-packet problem we've encountered, which L248 for example assumes cannot happen. Without considering that, then this will eventually read an out-of-bounds address, which will either cause process termination or silent data corruption.

Reply:
DJ will bring this up at the pass1 workshop.

Reply:
Following an offline discussion, I had misunderstood the nature of the errors we currently expect to encounter. We may go out-of-sequence when parsing the packed data, but there is not a mechanism by which we foresee encountering an out-of-bounds read like I suggested. Pasha will implement a simple check for having gone out-of-sequence, and utilize the `StrawDigiFlag` data product to mark digis in such an event as having been part of problematic event.